### ISSUE TITLE
`reportUnboundVariable` false positive in walrus operator inside for loop followed by assignment and another for loop

### ISSUE THREAD

**Issue by DetachHead** [DetachHead/basedpyright#642]
Code sample in [basedpyright playground](https://basedpyright.com/?pythonVersion=3.13&typeCheckingMode=all&reportUnknownArgumentType=false&reportArgumentType=false&code=CYUwZgBGD20BQCgLIgNwIYBsCuICMAXBAJYB2ALhAD4QBy0pIANGlrgExGbEDO5A2mXIBdBAEoCSFDABOEAPolSrHCE5SUKDKoCsEALwQADjKFw423ABYIBQ5fxiWAW3QAPC23wsHVsWI1NFVwAZgMIfgcdUSCUYkgHPCVgkBDJWNiAOmygA)

```python
def foo(
    value1: int | None, value2: list[int]
):
    for _ in value2:
        value5 = print((value4 := value1), max(value1, value4))
        value3 = [value5]
        if value1 in value3:
            ...
```

very odd bug. this is as minified as i could get it, though the maxified version seems to only happen in the language server but not the cli for some reason

**Comment by KotlinIsland** [comment]
we mypy now?
```py
def f1(a: object, b: object) -> object: ...

def f2[T](a: object, b: T) -> object: ...

value1 = 1
for _ in [1]:
    value3 = f1(value2 := value1, f2(value1, value2))
    if value1 == value3:
        ...
```

**Comment by DetachHead** [comment]
correction: it doesn't only happen in the language server, instead it only happens when typechecking the file but not the whole project
